cluster/job.go

```diff
@@ -27,17 +27,73 @@ type JobConfig struct {
 	Interval time.Duration
 }
 
+// NextWaitInterval is a callback computing the next wait interval for a job.
+type NextWaitInterval func(now time.Time, metadata jobMetadata) time.Duration
+
+// MakeWaitForInterval creates a function to scheduling a job to run on the given interval relative
+// to the last finished timestamp.
+//
+// For example, if the job first starts at 12:01 PM, and is configured with interval 5 minutes,
+// it will next run at:
+//     12:06, 12:11, 12:16, ...
+//
+// If the job has not previously started, it will run immediately.
+func MakeWaitForInterval(interval time.Duration) NextWaitInterval {
+	if interval == 0 {
+		panic("must specify non-zero ready interval")
+	}
+
+	return func(now time.Time, metadata jobMetadata) time.Duration {
+		sinceLastFinished := now.Sub(metadata.LastFinished)
+		if sinceLastFinished < interval {
+			return interval - sinceLastFinished
+		}
+
+		return 0
+	}
+}
+
+// MakeWaitForRoundedInterval creates a function, scheduling a job to run on the nearest rounded
+// interval relative to the last finished timestamp.
+//
+// For example, if the job first starts at 12:04 PM, and is configured with interval 5 minutes,
+// and is configured to round to 5 minute intervals, it will next run at:
+//	12:05 PM, 12:10 PM, 12:15 PM, ...
+//
+// If the job has not previously started, it will run immediately. Note that this wait interval
+// strategy does not guarantee a minimum interval between runs, only that subsequent runs will be
+// scheduled on the rounded interval.
+func MakeWaitForRoundedInterval(interval time.Duration) NextWaitInterval {
+	if interval == 0 {
+		panic("must specify non-zero ready interval")
+	}
+
+	return func(now time.Time, metadata jobMetadata) time.Duration {
+		if metadata.LastFinished.IsZero() {
+			return 0
+		}
+
+		target := metadata.LastFinished.Add(interval).Truncate(interval)
```

Comment:
Great way to handle this! If I understand this correctly, this is how an example situation would play out:

- Rounded interval is set to `5 minutes`
- Job starts at `12:04PM`
- First run runs at `12:05PM`
- The job lasts `7 minutes`
- The job will not run at `12:10PM`, and will instead run next at `12:15PM`

A job meant to run every 5 minutes should not last that long (of course it could), but I just want to make sure I understand the functionality.

Reply:
That's (almost) correct -- we anchor from the time the job last finished to determine when to next run the job. Since the job would have finished at `12:11PM`, it would indeed next run at `12:15PM`.

The nuance is that the "first run" will actually occur at `12:04PM`, with the second run occurring at `12:05PM` and subsequent runs then being properly spaced out. I opted for this edge case handling because it eliminated the need to supply a `startTime` from which to anchor the calculations. Without an explicit `startTime` (i.e. when there is no `metadata.LastFinished` yet), it's not obvious how to wait for the top of the interval. By the time we reach it, we're in a new interval, and might decide to wait for the top of /that/ one. 

I suppose we could tweak this so that `12:00:00-12:02:29` would round down and `12:02:30-12:04:49` would round up. My assumption was that these jobs would likely be "idempotent" and that running slightly more often that necessary in one case would be worth the simpler API surface and implementation.

Reply:
I like the current implementation's consistency, 1/5 against special-cased round-down